README
NAME
    Unicode::Block - Class for unicode block manipulation.

SYNOPSIS
     use Unicode::Block;

     my $obj = Unicode::Block->new(%parameters);
     my $item = $obj->next;

METHODS
  "new"
     my $obj = Unicode::Block->new(%parameters);

    Constructor.

    *       "char_from"

            Character from.

            Default value is '0000'.

    *       "char_to"

            Character to.

            Default value is '007f'.

    *       "title"

            Title of block.

            Default value is undef.

    Returns instance of object.

  "next"
     my $item = $obj->next;

    Get next character.

    Returns Unicode::Block::Item object for character, if character exists.
    Returns undef, if character doesn't exist.

ERRORS
     new():
             From Class::Utils::set_params_pub():
                     Unknown parameter '%s'.

EXAMPLE
     use strict;
     use warnings;

     use Unicode::Block;
     use Unicode::UTF8 qw(encode_utf8);

     # Object.
     my $obj = Unicode::Block->new;

     # Print all.
     my $num = 0;
     while (my $char = $obj->next) {
            if ($num != 0) {
                    if ($num % 16 == 0) {
                            print "\n";
                    } else {
                            print " ";
                    }
            }
            print encode_utf8($char->char);
            $num++;
     }
     print "\n";

     # Output.
     #                                
     #                                
     #   ! " # $ % & ' ( ) * + , - . /
     # 0 1 2 3 4 5 6 7 8 9 : ; < = > ?
     # @ A B C D E F G H I J K L M N O
     # P Q R S T U V W X Y Z [ \ ] ^ _
     # ` a b c d e f g h i j k l m n o
     # p q r s t u v w x y z { | } ~

DEPENDENCIES
    Class::Utils, Unicode::Block::Item.

SEE ALSO
    Unicode::Block::Ascii
        Ascii output of unicode block.

    Unicode::Block::List
        List of unicode blocks.

REPOSITORY
    <https://github.com/tupinek/Unicode-Block>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
    © 2013-2023 Michal Josef Špaček

    BSD 2-Clause License

VERSION
    0.09



Changes
0.09
 - Sort used modules in tests.

0.08 2023-11-21T14:40:37+01:00
 - Improve LICENSE AND COPYRIGHT section in doc.
 - Improve SYNOPSIS section in doc.
 - Improve constructor parameters in doc.
 - Move bugtracker to github.
 - Rename example files to better names.
 - Rewrite Encode::decode_utf8 to better Unicode::UTF8::decode_utf8.
 - Rewrite doc to use =head2 in methods.
 - Update Module::Install to 1.21 version.
 - Update author github username.
 - Update copyright years.

0.07 2018-02-07T07:54:14+01:00
 - Fix Makefile.PL for new version of Perl and no default »use lib '.'«.

0.06 2017-12-26T01:37:41+01:00
 - Fix Makefile.PL for new version of Perl and no default »use lib '.'«.
 - Improve LICENSE AND COPYRIGHT section in doc.
 - Improve SEE ALSO section in doc.

0.05 2017-12-25T16:29:54+01:00
 - Add information about hex_length in doc.
 - Fix running of author tests.
 - Fix test of Brahmi block in Perl lesser than 5.14.0.
 - Fix test comments.

0.04 2017-12-19T14:12:57+01:00
 - Add SEE ALSO section to Unicode::Block doc.
 - Add SEE ALSO section to Unicode::Block::Item doc.
 - Fix test 09-width.t for C locale.
 - Rewrite tests to author tests.
 - Rewrite unassigned unicode character to '' (blank space).
 - Update Module::Install to 1.18 version.
 - Update author name.
 - Update copyright years.

0.03 2014-08-26T03:04:50+02:00
 - Add minimal version of Perl.
 - Add tests.

0.02 2014-08-20T09:05:16+02:00
 - Fix dependency to Class::Utils 0.06 version.

0.01 2014-08-20T07:36:40+02:00
 - First version.


Block.pm
package Unicode::Block;

use strict;
use warnings;

use Class::Utils qw(set_params_pub);
use Unicode::Block::Item;

our $VERSION = 0.09;

# Constructor.
sub new {
	my ($class, @params) = @_;

	# Create object.
	my $self = bless {}, $class;

	# Character from.
	$self->{'char_from'} = '0000',

	# Character to.
	$self->{'char_to'} = '007F',

	# Title.
	$self->{'title'} = undef;

	# Process parameters.
	set_params_pub($self, @params);

	# Count.
	$self->{'_count'} = $self->{'char_from'};

	# Object.
	return $self;
}

# Get next character.
sub next {
	my $self = shift;
	my $char_hex = $self->_count;
	if (defined $char_hex) {
		return Unicode::Block::Item->new('hex' => $char_hex);
	} else {
		return;
	}
}

# Get actual character and increase number.
sub _count {
	my $self = shift;
	my $ret = $self->{'_count'};
	if (! defined $ret) {
		return;
	}
	my $num = hex $self->{'_count'};
	$num++;
	my $last_num = hex $self->{'char_to'};
	if ($num > $last_num) {
		$self->{'_count'} = undef;
	} else {
		$self->{'_count'} = sprintf '%x', $num;
	}
	return $ret;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

Unicode::Block - Class for unicode block manipulation.

=head1 SYNOPSIS

 use Unicode::Block;

 my $obj = Unicode::Block->new(%parameters);
 my $item = $obj->next;

=head1 METHODS

=head2 C<new>

 my $obj = Unicode::Block->new(%parameters);

Constructor.

=over 8

=item * C<char_from>

Character from.

Default value is '0000'.

=item * C<char_to>

Character to.

Default value is '007f'.

=item * C<title>

Title of block.

Default value is undef.

=back

Returns instance of object.

=head2 C<next>

 my $item = $obj->next;

Get next character.

Returns Unicode::Block::Item object for character, if character exists.
Returns undef, if character doesn't exist.

=head1 ERRORS

 new():
         From Class::Utils::set_params_pub():
                 Unknown parameter '%s'.

=head1 EXAMPLE

=for comment filename=print_ascii_block.pl

 use strict;
 use warnings;

 use Unicode::Block;
 use Unicode::UTF8 qw(encode_utf8);

 # Object.
 my $obj = Unicode::Block->new;

 # Print all.
 my $num = 0;
 while (my $char = $obj->next) {
        if ($num != 0) {
                if ($num % 16 == 0) {
                        print "\n";
                } else {
                        print " ";
                }
        }
        print encode_utf8($char->char);
        $num++;
 }
 print "\n";

 # Output.
 #                                
 #                                
 #   ! " # $ % & ' ( ) * + , - . /
 # 0 1 2 3 4 5 6 7 8 9 : ; < = > ?
 # @ A B C D E F G H I J K L M N O
 # P Q R S T U V W X Y Z [ \ ] ^ _
 # ` a b c d e f g h i j k l m n o
 # p q r s t u v w x y z { | } ~  

=head1 DEPENDENCIES

L<Class::Utils>,
L<Unicode::Block::Item>.

=head1 SEE ALSO

=over

=item L<Unicode::Block::Ascii>

Ascii output of unicode block.

=item L<Unicode::Block::List>

List of unicode blocks.

=back

=head1 REPOSITORY

L<https://github.com/tupinek/Unicode-Block>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2013-2023 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.09

=cut


Block/Item.pm
package Unicode::Block::Item;

use strict;
use warnings;

use Class::Utils qw(set_params);
use Error::Pure qw(err);
use Readonly;
use Text::CharWidth qw(mbwidth);
use Unicode::Char;

# Constants.
Readonly::Scalar our $EMPTY_STR => q{};
Readonly::Scalar our $SPACE => q{ };

our $VERSION = 0.09;

# Constructor.
sub new {
	my ($class, @params) = @_;

	# Create object.
	my $self = bless {}, $class;

	# Hexadecimal number.
	$self->{'hex'} = undef;

	# Length of hex number.
	$self->{'hex_length'} = 4;

	# Process parameters.
	set_params($self, @params);

	# Check hex number.
	if (! defined $self->{'hex'}) {
		err "Parameter 'hex' is required.";
	}
	if (! $self->_is_hex) {
		err "Parameter 'hex' isn't hexadecimal number.";
	}

	# Object.
	return $self;
}

# Get hex base number.
sub base {
	my $self = shift;
	my $base = substr $self->hex, 0, -1;
	return 'U+'.$base.'x';
}

# Get character.
sub char {
	my $self = shift;

	# Create object.
	if (! exists $self->{'u'}) {
		$self->{'u'} = Unicode::Char->new;
	}

	# Get char.
	my $char = $self->{'u'}->u($self->{'hex'});

	# 'Non-Spacing Mark' and 'Enclosing Mark'.
	if ($char =~ m/\p{Mn}/ms || $char =~ m/\p{Me}/ms) {
		$char = $SPACE.$char;

	# Control.
	} elsif ($char =~ m/\p{Cc}/ms) {
		$char = $SPACE;

	# Not Assigned.
	} elsif ($char =~ m/\p{Cn}/ms) {
		$char = $EMPTY_STR;
	}

	return $char;
}

# Get hex number.
sub hex {
	my $self = shift;
	return sprintf '%0'.$self->{'hex_length'}.'x',
		CORE::hex $self->{'hex'};
}

# Get last hex number.
sub last_hex {
	my $self = shift;
	return substr $self->{'hex'}, -1;
}

# Get character width.
sub width {
	my $self = shift;
	if (! exists $self->{'_width'}) {
		$self->{'_width'} = mbwidth($self->char);
		if ($self->{'_width'} == -1) {
			$self->{'_width'} = 1;
		}
	}
	return $self->{'_width'};
}

# Check for hex number.
sub _is_hex {
	my $self = shift;
	if ($self->{'hex'} !~ m/^[0-9a-fA-F]+$/ims) {
		return 0;
	}
	my $int = CORE::hex $self->{'hex'};
	if (! defined $int) {
		return 0;
	}
	my $hex = sprintf '%x', $int;
	my $value = lc $self->{'hex'};
	$value =~ s/^0*//ms;
	if ($value eq '') {
		$value = 0;
	}
	if ($hex ne $value) {
		return 0;
	}
	return 1;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

Unicode::Block::Item - Class for unicode block character.

=head1 SYNOPSIS

 use Unicode::Block::Item;

 my $obj = Unicode::Block::Item->new(%parameters);
 my $base = $obj->base;
 my $char = $obj->char;
 my $hex = $obj->hex;
 my $last_hex = $obj->last_hex;
 my $width = $obj->width;

=head1 METHODS

=head2 C<new>

 my $obj = Unicode::Block::Item->new(%parameters);

Constructor.

=over 8

=item * C<hex>

Hexadecimal number.

It is required.

Default value is undef.

=item * C<hex_length>

Length of hex number.
It's used for formatting of hex() method output.

Default value is 4.

=back

Returns instance of object.

=head2 C<base>

 my $base = $obj->base;

Get hex base number in format 'U+???x'.
Example: 'hex' => 1234h; Returns 'U+123x'.

Returns string with hex base number.

=head2 C<char>

 my $char = $obj->char;

Get character.
Example: 'hex' => 1234h; Returns 'ሴ'.

Returns string with character.

=head2 C<hex>

 my $hex = $obj->hex;

Get hex number in 'hex_length' length.
Example: 'hex' => 1234h; Returns '0x1234'.

Returns string with hex number.

=head2 C<last_hex>

 my $last_hex = $obj->last_hex;

Get last hex number.
Example: 'hex' => 1234h; Returns '4'.

Returns string with last hex number.

=head2 C<width>

 my $width = $obj->width;

Get character width.

Returns string with width.

=head1 ERRORS

 new():
         Parameter 'hex' is required.
         Parameter 'hex' isn't hexadecimal number.
         From Class::Utils::set_params():
                 Unknown parameter '%s'.

=head1 EXAMPLE

=for comment filename=print_info_about_example_character.pl

 use strict;
 use warnings;

 use Unicode::Block::Item;
 use Unicode::UTF8 qw(encode_utf8);

 # Object.
 my $obj = Unicode::Block::Item->new(
        'hex' => 2505,
 );

 # Print out.
 print 'Character: '.encode_utf8($obj->char)."\n";
 print 'Hex: '.$obj->hex."\n";
 print 'Last hex character: '.$obj->last_hex."\n";
 print 'Base: '.$obj->base."\n";

 # Output.
 # Character: ┅
 # Hex: 2505
 # Last hex character: 5
 # Base: U+250x

=head1 DEPENDENCIES

L<Class::Utils>,
L<Error::Pure>,
L<Readonly>,
L<Text::CharWidth>,
L<Unicode::Char>.

=head1 SEE ALSO

=over

=item L<Unicode::Block>

Class for unicode block manipulation.

=back

=head1 REPOSITORY

L<https://github.com/tupinek/Unicode-Block>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2013-2023 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.09

=cut


Makefile.PL
use lib '.';
use strict;
use warnings;

use inc::Module::Install;

# Definition.
abstract 'Class for unicode block manipulation.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.05;
author_requires 'Test::More' => 0;
author_requires 'Test::NoWarnings' => 0;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
license 'bsd';
name 'Unicode-Block';
readme_from 'Block.pm';
recursive_author_tests('xt');
requires 'Class::Utils' => 0.06;
requires 'Error::Pure' => 0.13;
requires 'Text::CharWidth' => 0;
requires 'Unicode::Char' => 0;
requires 'perl' => '5.6.2';
resources 'bugtracker' => 'https://github.com/michal-josef-spacek/Unicode-Block/issues';
resources 'homepage' => 'https://github.com/michal-josef-spacek/Unicode-Block';
resources 'repository' => 'git://github.com/michal-josef-spacek/Unicode-Block';
test_requires 'English' => 0;
test_requires 'Error::Pure::Utils' => 0.13;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
test_requires 'Unicode::UTF8' => 0;
tests_recursive;
version '0.09';

# Run.
WriteAll();
